// sphere-node-sdk/lib/sync/utils/order.js
var BaseUtils, OrderUtils, _, actionsList, actionsListReturnInfoState,
  extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty;

_ = require('underscore');

BaseUtils = require('./base');

OrderUtils = (function(superClass) {
  extend(OrderUtils, superClass);

  function OrderUtils() {
    return OrderUtils.__super__.constructor.apply(this, arguments);
  }

  OrderUtils.prototype.diff = function(old_obj, new_obj) {
    var patchReturnInfos;
    patchReturnInfos = function(order) {
      return _.each(order.returnInfo, function(info, index) {
        return info._MATCH_CRITERIA = "" + index;
      });
    };
    patchReturnInfos(old_obj);
    patchReturnInfos(new_obj);
    return OrderUtils.__super__.diff.call(this, old_obj, new_obj);
  };

  OrderUtils.prototype.actionsMapStatusValues = function(diff, old_obj) {
    var actions;
    actions = [];
    _.each(actionsList(), (function(_this) {
      return function(item) {
        var action, key, obj, updated;
        key = item.key;
        obj = diff[key];
        if (obj) {
          updated = _this.getDeltaValue(obj);
          action = {
            action: item.action
          };
          action[key] = updated;
        }
        if (action) {
          return actions.push(action);
        }
      };
    })(this));
    return actions;
  };

  OrderUtils.prototype.actionsMapDeliveries = function(diff, old_obj) {
    var actions;
    if (!(_.has(diff, 'shippingInfo') && _.has(diff.shippingInfo, 'deliveries'))) {
      return [];
    }
    actions = _.chain(diff.shippingInfo.deliveries).filter(function(item, key) {
      return key !== '_t';
    }).map(function(deliveryDiff, deliveryIndex) {
      var action, delivery;
      if (_.isArray(deliveryDiff)) {
        delivery = _.last(deliveryDiff);
        action = {
          action: 'addDelivery'
        };
        _.each(_.keys(delivery), function(key) {
          return action[key] = delivery[key];
        });
        return action;
      } else {
        return _.chain(deliveryDiff.parcels).filter(function(item, key) {
          return key !== '_t' && _.isArray(item);
        }).map(function(parcelDiff) {
          var parcel;
          parcel = _.last(parcelDiff);
          action = {
            action: 'addParcelToDelivery',
            deliveryId: old_obj.shippingInfo.deliveries[deliveryIndex].id
          };
          _.each(parcel, function(item, key) {
            return action[key] = item;
          });
          return action;
        }).value();
      }
    }).value();
    return _.flatten(actions);
  };

  OrderUtils.prototype.actionsMapReturnInfo = function(diff, old_obj) {
    var actions;
    if (!_.has(diff, 'returnInfo')) {
      return [];
    }
    actions = _.chain(diff['returnInfo']).filter(function(item, key) {
      return key !== '_t';
    }).map((function(_this) {
      return function(returnInfoDelta, returnInfoDeltaKey) {
        var action, returnInfo;
        if (_.isArray(returnInfoDelta)) {
          returnInfo = _.last(returnInfoDelta);
          action = {
            action: 'addReturnInfo'
          };
          _.each(returnInfo, function(value, key) {
            return action[key] = value;
          });
          return action;
        } else {
          returnInfo = returnInfoDelta;
          return actions = _.chain(returnInfo.items).filter(function(item, key) {
            return key !== '_t';
          }).map(function(item, itemKey) {
            return _.chain(actionsListReturnInfoState()).filter(function(actionDefinition) {
              return _.has(item, actionDefinition.key);
            }).map(function(actionDefinition) {
              action = {
                action: actionDefinition.action,
                returnItemId: old_obj.returnInfo[returnInfoDeltaKey].items[itemKey].id
              };
              action[actionDefinition.key] = _this.getDeltaValue(item[actionDefinition.key]);
              return action;
            }).value();
          }).value();
        }
      };
    })(this)).value();
    return _.flatten(actions);
  };

  return OrderUtils;

})(BaseUtils);

module.exports = OrderUtils;

actionsList = function() {
  return [
    {
      action: 'changeOrderState',
      key: 'orderState'
    }, {
      action: 'changePaymentState',
      key: 'paymentState'
    }, {
      action: 'changeShipmentState',
      key: 'shipmentState'
    }
  ];
};

actionsListReturnInfoState = function() {
  return [
    {
      action: 'setReturnShipmentState',
      key: 'shipmentState'
    }, {
      action: 'setReturnPaymentState',
      key: 'paymentState'
    }
  ];
};
