// sphere-node-client/lib/client.js
/* ===========================================================
# sphere-node-client - v0.8.0
# ==============================================================
# Copyright (c) 2014 SPHERE.IO™
# Licensed MIT.
#
#    https://github.com/sphereio/sphere-node-client/blob/master/LICENSE-MIT
*/
var CartService, CategoryService, ChannelService, CommentService, CustomObjectService, CustomerGroupService, CustomerService, InventoryEntryService, Logger, MessageService, OrderService, ProductDiscountService, ProductProjectionService, ProductService, ProductTypeService, Rest, ReviewService, ShippingMethodService, SphereClient, StateService, TaskQueue, TaxCategoryService, ZoneService, _;

_ = require('underscore');

Rest = require('sphere-node-connect').Rest;

TaskQueue = require('sphere-node-utils').TaskQueue;

Logger = require('./logger');

CartService = require('./services/carts');

CategoryService = require('./services/categories');

ChannelService = require('./services/channels');

CommentService = require('./services/comments');

CustomObjectService = require('./services/custom-objects');

CustomerService = require('./services/customers');

CustomerGroupService = require('./services/customer-groups');

InventoryEntryService = require('./services/inventory-entries');

MessageService = require('./services/messages');

OrderService = require('./services/orders');

ProductDiscountService = require('./services/product-discounts');

ProductService = require('./services/products');

ProductProjectionService = require('./services/product-projections');

ProductTypeService = require('./services/product-types');

ReviewService = require('./services/reviews');

ShippingMethodService = require('./services/shipping-methods');

StateService = require('./services/states');

TaxCategoryService = require('./services/tax-categories');

ZoneService = require('./services/zones');


/**
 * Defines a SphereClient.
 * @class SphereClient
 */

SphereClient = (function() {

  /**
   * Constructs a new client with given API credentials
   * @constructor
   *
   * @param {Object} [options] An object containing the credentials for the `sphere-node-connect`
   * {@link https://github.com/sphereio/sphere-node-connect#documentation}
   */
  function SphereClient(options) {
    var _serviceOptions;
    if (options == null) {
      options = {};
    }

    /**
     * @private
     * Instance of a Logger
     * @type {Logger}
     */
    this._logger = new Logger(options.logConfig);

    /**
     * @private
     * Instance of a TaskQueue
     * @type {TaskQueue}
     */
    this._task = options.task || new TaskQueue;

    /**
     * @private
     * Instance of the Rest client
     * @type {Rest}
     */
    this._rest = options.rest || new Rest(_.extend(options, {
      logConfig: {
        logger: this._logger
      }
    }));

    /**
     * @private
     * Wrapper to pass different options to new service instances
     * @type {Object}
     */
    _serviceOptions = {
      _rest: this._rest,
      _task: this._task,
      _logger: this._logger,
      _stats: _.defaults(options.stats || {}, {
        includeHeaders: false
      })
    };
    this.carts = new CartService(_serviceOptions);
    this.categories = new CategoryService(_serviceOptions);
    this.channels = new ChannelService(_serviceOptions);
    this.comments = new CommentService(_serviceOptions);
    this.customObjects = new CustomObjectService(_serviceOptions);
    this.customers = new CustomerService(_serviceOptions);
    this.customerGroups = new CustomerGroupService(_serviceOptions);
    this.inventoryEntries = new InventoryEntryService(_serviceOptions);
    this.messages = new MessageService(_serviceOptions);
    this.orders = new OrderService(_serviceOptions);
    this.products = new ProductService(_serviceOptions);
    this.productDiscounts = new ProductDiscountService(_serviceOptions);
    this.productProjections = new ProductProjectionService(_serviceOptions);
    this.productTypes = new ProductTypeService(_serviceOptions);
    this.reviews = new ReviewService(_serviceOptions);
    this.shippingMethods = new ShippingMethodService(_serviceOptions);
    this.states = new StateService(_serviceOptions);
    this.taxCategories = new TaxCategoryService(_serviceOptions);
    this.zones = new ZoneService(_serviceOptions);
  }


  /**
   * Define max parallel request to be sent on each request from the {TaskQueue}
   * @param {Number} maxParallel A number between 1 and 100 (default is 20)
   */

  SphereClient.prototype.setMaxParallel = function(maxParallel) {
    return this._task.setMaxParallel(maxParallel);
  };

  return SphereClient;

})();


/**
 * The {@link SphereClient} client.
 */

module.exports = SphereClient;
