// sphere-node-sdk/lib/sync/base-sync.js
var BaseSync, BaseUtils, _, debug;

debug = require('debug')('sphere-sync');

_ = require('underscore');

BaseUtils = require('./utils/base');

BaseSync = (function() {
  function BaseSync() {
    this._data = {};
    this._utils = new BaseUtils;
    this._syncConfig = [];
  }

  BaseSync.prototype.config = function(actionGroups) {
    this._syncConfig = actionGroups || [];
    return this;
  };

  BaseSync.prototype.buildActions = function(new_obj, old_obj) {
    var actions, diff, update;
    if (!new_obj || !old_obj) {
      throw new Error('Missing either new_obj or old_obj in order to build update actions');
    }
    diff = this._utils.diff(old_obj, new_obj);
    debug('JSON diff for %s Sync: %j', this.constructor.name, diff);
    update = void 0;
    if (diff) {
      actions = this._doMapActions(diff, new_obj, old_obj);
      if (actions.length > 0) {
        update = {
          actions: actions,
          version: old_obj.version
        };
      }
    }
    this._data = {
      update: update,
      updateId: old_obj.id
    };
    debug('JSON data update for %s Sync: %j', this.constructor.name, this._data);
    return this;
  };

  BaseSync.prototype.filterActions = function(fn) {
    var filtered;
    if (!fn) {
      return this;
    }
    if (!this._data.update) {
      return this;
    }
    filtered = _.filter(this._data.update.actions, fn);
    if (_.isEmpty(filtered)) {
      this._data.update = void 0;
    } else {
      this._data.update.actions = filtered;
    }
    return this;
  };

  BaseSync.prototype.shouldUpdate = function() {
    return !_.isEmpty(this._data.update);
  };

  BaseSync.prototype.getUpdateId = function() {
    var ref;
    return (ref = this._data) != null ? ref.updateId : void 0;
  };

  BaseSync.prototype.getUpdateActions = function() {
    var ref, ref1;
    return ((ref = this._data) != null ? (ref1 = ref.update) != null ? ref1.actions : void 0 : void 0) || [];
  };

  BaseSync.prototype.getUpdatePayload = function() {
    var ref;
    return (ref = this._data) != null ? ref.update : void 0;
  };

  BaseSync.prototype._mapActionOrNot = function(type, fn) {
    var found;
    if (_.isEmpty(this._syncConfig)) {
      return fn();
    }
    found = _.find(this._syncConfig, function(c) {
      return c.type === type;
    });
    if (!found) {
      return [];
    }
    switch (found.group) {
      case 'black':
        return [];
      case 'white':
        return fn();
      default:
        throw new Error("Action group '" + found.group + "' not supported. Please use black or white.");
    }
  };

  BaseSync.prototype._doMapActions = function(diff, new_obj, old_obj) {
    return [];
  };

  return BaseSync;

})();

module.exports = BaseSync;
